### PR TITLE
feat: refactor reconciler svc

### PR DESCRIPTION
**Pull request by mfiedorowicz**
- extract methods for generating and applying change sets into dedicated packages (differ, applier)
- generate and apply change sets concurrently, including configurable rate limiter to control pressure on NetBox (as it's capabilities depends on available resource limits (CPU, memory, postgres, etc.)

Integration tests:
Screenshot 2024-11-20 at 11 26 12

### CODE REVIEW

File: diode-server/reconciler/ingestion_processor.go [hunk 1]
```diff
@@ -186,6 +197,188 @@ func (p *IngestionProcessor) handleStreamMessage(ctx context.Context, msg redis.
...
+		for ingestionLog := range applyChan {
+			if err := limiter.Wait(ctx); err != nil {
+				p.logger.Debug("rate limiter wait", "error", err)
+				return
+			}
+
+			p.logger.Debug("applying changeset", "ingestionLogID", ingestionLog.ingestionLog.GetId(), "changeSetID", ingestionLog.changeSet.ChangeSetID)
+
+			if err := applier.ApplyChangeSet(ctx, p.logger, *ingestionLog.changeSet, p.nbClient); err != nil {
+				ingestionLog.errors = append(ingestionLog.errors, fmt.Errorf("failed to apply change set: %v", err))
```
Comment (leoparente) [inline diode-server/reconciler/ingestion_processor.go]:
```suggestion
				ingestionLog.errors = append(ingestionLog.errors, fmt.Errorf("failed to apply changeset: %v", err))
```

File: diode-server/reconciler/ingestion_processor_internal_test.go [hunk 2]
```diff
@@ -81,6 +83,11 @@ func TestWriteIngestionLog(t *testing.T) {
 			p := &IngestionProcessor{
 				redisClient: mockRedisClient,
 				logger:      slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelDebug, AddSource: false})),
+				Config: Config{
+					AutoApplyChangesets:        true,
```
Comment (leoparente) [inline diode-server/reconciler/ingestion_processor_internal_test.go:87]: Should we add a test with apply false to check outcome?

Comment (mfiedorowicz) [reply diode-server/reconciler/ingestion_processor_internal_test.go:87]: Good point! Thank you

**Review by leoparente — APPROVED** [review]
Great work! LGTM

**Pull request merged**